File: backend/app/models/artifact.py
"""Artifact catalog models (docs/12).

An *artifact* is a real-world thing a video references — a book, movie, podcast,
product, place, etc. — as opposed to the structured how-to content of the card
itself. Artifacts are extracted by the same single structuring LLM call (docs/04),
then aggregated across all cards into a global, deduplicated catalog.

This is a parallel surface to the block schema, NOT a new block type. The block
vocabulary is unchanged; the structuring output simply grows an `artifacts` list.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class ArtifactType(str, Enum):
    BOOK = "book"
    MOVIE = "movie"
    TV_SHOW = "tv_show"
    PODCAST = "podcast"
    MUSIC = "music"
    PRODUCT = "product"
    PLACE = "place"
    APP = "app"
    OTHER = "other"


def _new_id() -> str:
    return "a_" + uuid.uuid4().hex[:8]


class Artifact(BaseModel):
    """One referenced thing, as emitted by structuring (pre-aggregation)."""

    type: ArtifactType = ArtifactType.OTHER
    title: str
    creator: Optional[str] = None  # author / director / artist / host
    year: Optional[int] = None
    thumbnail: Optional[str] = None  # resolved from a free image API (docs/12)


class CatalogEntry(BaseModel):
    """An aggregated, deduplicated catalog item: one artifact, many source cards."""

    id: str = Field(default_factory=_new_id)
    type: ArtifactType = ArtifactType.OTHER
    title: str
    creator: Optional[str] = None
    year: Optional[int] = None
    thumbnail: Optional[str] = None
    source_card_ids: list[str] = Field(default_factory=list)
    created_at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    # User must explicitly save a referenced artifact into the catalog (long-press);
    # unsaved rows still exist as per-card references but stay out of the catalog tab.
    saved: bool = False
    # Optional LLM-generated "what is this" detail, filled on-demand via Fetch info.
    description: Optional[str] = None

File: app/lib/domain/models/artifact.dart
/// Catalog artifact: a real-world thing a video references — book, movie,
/// podcast, product, place, etc. Mirrors the backend catalog contract
/// (docs/12, backend models/artifact.py). Parsing is tolerant: unknown type
/// degrades to [ArtifactType.other] rather than throwing.
library;

enum ArtifactType {
  book,
  movie,
  tvShow,
  podcast,
  music,
  product,
  place,
  app,
  other;

  static ArtifactType fromWire(String? value) {
    switch (value) {
      case 'book':
        return ArtifactType.book;
      case 'movie':
        return ArtifactType.movie;
      case 'tv_show':
        return ArtifactType.tvShow;
      case 'podcast':
        return ArtifactType.podcast;
      case 'music':
        return ArtifactType.music;
      case 'product':
        return ArtifactType.product;
      case 'place':
        return ArtifactType.place;
      case 'app':
        return ArtifactType.app;
      default:
        return ArtifactType.other;
    }
  }

  String get wire {
    switch (this) {
      case ArtifactType.tvShow:
        return 'tv_show';
      default:
        return name;
    }
  }

  /// Plural label used as a catalog section header.
  String get sectionLabel {
    switch (this) {
      case ArtifactType.book:
        return 'Books';
      case ArtifactType.movie:
        return 'Movies';
      case ArtifactType.tvShow:
        return 'TV Shows';
      case ArtifactType.podcast:
        return 'Podcasts';
      case ArtifactType.music:
        return 'Music';
      case ArtifactType.product:
        return 'Products';
      case ArtifactType.place:
        return 'Places';
      case ArtifactType.app:
        return 'Apps';
      case ArtifactType.other:
        return 'Other';
    }
  }
}

class CatalogEntry {
  const CatalogEntry({
    required this.id,
    this.type = ArtifactType.other,
    required this.title,
    this.creator,
    this.year,
    this.thumbnail,
    this.sourceCardIds = const [],
    this.saved = false,
    this.description,
  });

  final String id;
  final ArtifactType type;
  final String title;
  final String? creator;
  final int? year;
  final String? thumbnail;
  final List<String> sourceCardIds;

  /// Whether the user saved this into the catalog tab (long-press to save).
  final bool saved;

  /// On-demand LLM detail ("what it's about"); null until Fetch info is pressed.
  final String? description;

  /// "James Clear · 2018", "2018", or "" — the dimmed subtitle line.
  String get subtitle {
    final parts = <String>[
      if (creator != null && creator!.isNotEmpty) creator!,
      if (year != null) '$year',
    ];
    return parts.join(' · ');
  }

  factory CatalogEntry.fromJson(Map<String, dynamic> json) => CatalogEntry(
        id: (json['id'] as String?) ?? '',
        type: ArtifactType.fromWire(json['type'] as String?),
        title: (json['title'] as String?) ?? '',
        creator: json['creator'] as String?,
        year: (json['year'] as num?)?.toInt(),
        thumbnail: json['thumbnail'] as String?,
        sourceCardIds: (json['source_card_ids'] as List?)
                ?.map((e) => e.toString())
                .toList() ??
            const [],
        saved: (json['saved'] as bool?) ?? false,
        description: json['description'] as String?,
      );
}
